src/time_tracker/invoices/handlers.clj

```diff
@@ -25,18 +21,61 @@
        (util/normalize-entities)))
 
 (defn- get-timers-to-invoice
-  [connection start end projects]
+  [connection start end invoice-project?]
   (->> (timers-db/retrieve-between connection start end)
-       (filter #(projects (:project-id %)))
+       (filter #(invoice-project? (:project-id %)))
        (util/normalize-entities)))
 
+(defn- bigdecify-rates
+  [raw-rates-vector]
+  (map #(update % :rate  util/round-to-two-places) raw-rates-vector))
+
+(defn- bigdecify-taxes
+  "Converts tax-percentage to bigdec and returns nil if argument is nil."
+  [raw-taxes-vector]
+  (->> raw-taxes-vector
+       (map #(update % :tax-percentage util/round-to-two-places))
+       not-empty))
+
+(defn- coerce-and-validate-generate-invoice-body
+  "Coerces rates and tax percentages to bigdec and currency strings to keywords, and validates data."
+  [request-body]
+  ;; Validate the un-coerced data
+  (util/validate-spec request-body ::handlers-spec/generate-invoice-params-raw)
+  (let [coerced-body (-> request-body
+                         (update :user-id->rate bigdecify-rates)
+                         (update :tax-rates bigdecify-taxes)
+                         (update :currency keyword))]
+    (util/validate-spec coerced-body ::handlers-spec/generate-invoice-params-coerced)
+    coerced-body))
+
+(defn- id->name
+  [normalized-entities]
+  (fmap :name normalized-entities))
```

Comment:
There seem to be quite a few of these `->` fn names. Sometimes it feels like a good name because it describes what goes in and comes out. But, it's almost always better to describe what it _does_. `names`?